### DISCUSSION TITLE
Workflow Tools: How to use "itemSelector" ?

### DISCUSSION THREAD

**Discussion by grobert-gh** [nxt-dev/nxt_editor#250] [Q&A] [answered]
Hello everyone!
First of all, thank you for all your hardwork developing and sharing NXT. It is really a great tool and it is helping me a lot to automate and organize some daily tasks.

Recently I have been using Workflow Tools to build a small UI with rigging routines in Maya and so far it is working pretty well. But one thing I quite didn't figured out yet is how to use the "itemSelector" node.
When parenting it under a Button node it correctly appears as a menu item after right-clicking over that button in WTools interface. But the Item Selector window itself is always "empty" no matter what items are in `item_input_list` attribute. 

![Screen Shot 2022-08-06 at 21 57 29](https://user-images.githubusercontent.com/107314820/183249980-fe335c6e-9b9e-4230-9084-af0f8373c892.png)

![Screen Shot 2022-08-06 at 22 04 04](https://user-images.githubusercontent.com/107314820/183250074-64fbde88-eb46-4467-b492-cda23e2aa139.png)

Clicking "Accept" button returns the error bellow.

```
Traceback (most recent call last):
  File "/Users/User/nxt_maya/scripts/nxt_editor/dockwidgets/widget_builder.py", line 818, in <lambda>
    self.selection_widget(p, t))
  File "/Users/User/nxt_maya/scripts/nxt_editor/dockwidgets/widget_builder.py", line 886, in selection_widget
    target_node, target_attr = list_path_attr.split('.')
ValueError: need more than 1 value to unpack
```

At work I am currently project-locked to Maya 2020, so mostly using it with Python 2.7 daily. From a miniconda environment using Python 3.9, same error returns.

```
Traceback (most recent call last):
  File "/opt/miniconda3/envs/nxt/lib/python3.9/site-packages/nxt_editor/dockwidgets/widget_builder.py", line 818, in <lambda>
    self.selection_widget(p, t))
  File "/opt/miniconda3/envs/nxt/lib/python3.9/site-packages/nxt_editor/dockwidgets/widget_builder.py", line 886, in selection_widget
    target_node, target_attr = list_path_attr.split('.')
ValueError: not enough values to unpack (expected 2, got 1)
```

I know it is not documented yet but, there is a sample somewhere showing how it should be used?

Thank you in advance!
Cheers!

**Comment by ImLucasBrown** [comment, marked as answer]
Hi @grobert-gh,
Glad to hear NXT is adding value to your systems! Looks like you did stumble on a minor bug, NXT should be handling that error gracefully. To be totally fair, you're the first user I know of who's actually attempted to use this particular feature. 

I think the best way to describe the `itemSelector` is a way for you to pull information from your DCC into your graph while you're working. I've attached a really simple demo graph and gif of how it works ([itemSelector_demo.zip](https://github.com/nxt-dev/nxt_editor/files/9277138/itemSelector_demo.zip)). I can imagine a use case where you want to collect a list of nodes, lets say skinned geometry. You have another node that exports the skin cluster from a list of meshes. Executing the `/list_skinned_geo` node populates the `.meshes` attr on the `/export_skin_weights` node. This allows a sort of interactive editing of the graph. Just note that when you click `Accept` in the popup, you're actually setting a value on the graph, the selected items are set the same as if you had typed them into the target attr yourself.

Hopefully this is helpful!

![itemSelector_demo](https://user-images.githubusercontent.com/54835354/183300769-bc9bcd1e-7bc8-4117-a767-3790abd18b6a.gif)

**Reply by MrAaronAdams** [reply]
Alternatively, the itemSelector node itself could contain the "scene scraping" code that populates the `item_input_list`.  The Accept button updates the attribute found at this path defined on the itemSelector node: `item_list_attr_path`.  This path would likely be an attribute on the parent button node just above the itemSelector node.  That way one could simply click the same button after setting the list to work with.

![image](https://user-images.githubusercontent.com/18148685/183322469-8f5869f7-5318-4283-8411-ce115396d196.png)
![image](https://user-images.githubusercontent.com/18148685/183322384-e2d8e6c4-0273-4097-8cdc-9c56fb6b4c53.png)
![image](https://user-images.githubusercontent.com/18148685/183322403-dafdb0a4-b37d-433c-8b5a-8dbf57aa796d.png)
![image](https://user-images.githubusercontent.com/18148685/183322429-f0dc3c34-804d-44ad-b5ca-3104243e463d.png)

**Reply by grobert-gh** [reply]
Wonderfull!
Thank you both @ImLucasBrown and @MrAaronAdams !

It is very clear to me now, thanks to your samples and described usage. It opens more possibilities of interaction right on the UI, without having to update in Editor manually.

At first moment I was getting an empty popup but then I saw you unchecked "Workflow Tools always recomp" option and that did the trick. It's working like expected.

Thank you very much! ✌🏾
